File: umbrella-insurance-react-ui/src/models/users/policies/pendingPolicies/v1/PendingPolicy.ts
import { OrderType } from "../../../../orderTypes/v1/OrderType";
import { Peril } from "../../../../perils/v1/Peril";
import { Unit } from "../../../../units/v1/Unit";
import { AccountBalanceTransaction } from "../../../accountBalances/accountBalanceTransactions/v1/AccountBalanceTransaction";
import { Fee } from "../../../fees/v1/Fee";
import { Session } from "../../../sessions/v1/Session";
import { PendingPolicyState } from "../pendingPolicyStates/v1/PendingPolicyState";
import { PendingPolicyType } from "../pendingPolicyTypes/v1/PendingPolicyType";
import {Location} from "../../../../geographies/locations/v1/Location"

export class PendingPolicy {
    id?: number;
    session?: Session;
    premiumAmount?: number;
    startDate?: string;
    endDate?: string;
    coverageAmount?: number;
    impliedProbability?: number;
    unit?: Unit;//currency
    peril?: Peril;
    location?: Location;
    pendingPolicyState?: PendingPolicyState;
    fee?: Fee;
    pendingPolicyName?: string;
    splitPendingPolicy1?: PendingPolicy;
    splitPendingPolicy2?: PendingPolicy;
    originalPendingPolicy?: PendingPolicy;
    accountBalanceEscrowTransaction?: AccountBalanceTransaction;
    accountBalanceCanceledEscrowTransaction?: AccountBalanceTransaction;
    createdDateTime?: string
    canceledDateTime?: string;
    pendingPolicyType?: PendingPolicyType;
    orderType?: OrderType;

    constructor(obj?:any) {
        if(obj) {
            if(obj?.id) {
                this.id = obj.id;
            }
            if(obj?.session) {
                this.session = new Session(obj.session);
            }
            if(obj?.premiumAmount) {
                this.premiumAmount = obj.premiumAmount;
            }
            if(obj?.startDate) {
                this.startDate = obj.startDate;
            }
            if(obj?.endDate) {
                this.endDate = obj.endDate;
            }
            if(obj?.coverageAmount) {
                this.coverageAmount = obj.coverageAmount;
            }
            if(obj?.impliedProbability) {
                this.impliedProbability = obj.impliedProbability;
            }
            if(obj?.unit) {
                this.unit = new Unit(obj.unit);
            }
            if(obj?.peril) {
                this.peril = new Peril(obj.peril);
            }
            if(obj?.location) {
                this.location = new Location(obj.location);
            }
            if(obj?.pendingPolicyState) {
                this.pendingPolicyState = new PendingPolicyState(obj.pendingPolicyState);
            }
            if(obj?.fee) {
                this.fee = new Fee(obj.fee);
            }
            if(obj?.pendingPolicyName) {
                this.pendingPolicyName = obj.pendingPolicyName;
            }
            if(obj?.splitPendingPolicy1) {
                this.splitPendingPolicy1 = new PendingPolicy(obj.splitPendingPolicy1);
            }
            if(obj?.splitPendingPolicy2) {
                this.splitPendingPolicy2 = new PendingPolicy(obj.splitPendingPolicy2);
            }
            if(obj?.originalPendingPolicy) {
                this.originalPendingPolicy = new PendingPolicy(obj.originalPendingPolicy);
            }
            if(obj?.accountBalanceEscrowTransaction) {
                this.accountBalanceEscrowTransaction = new AccountBalanceTransaction(obj.accountBalanceEscrowTransaction);
            }
            if(obj?.accountBalanceCanceledEscrowTransaction) {
                this.accountBalanceCanceledEscrowTransaction = new AccountBalanceTransaction(obj.accountBalanceCanceledEscrowTransaction);
            }
            if(obj?.createdDateTime) {
                this.createdDateTime = obj.createdDateTime;
            }
            if(obj?.canceledDateTime) {
                this.canceledDateTime = obj.canceledDateTime;
            }
            if(obj?.pendingPolicyType) {
                this.pendingPolicyType = new PendingPolicyType(obj.pendingPolicyType);
            }
            if(obj?.orderType) {
                this.orderType = new OrderType(obj.orderType);
            }
        }
    }
}

File: umbrella-insurance-core/src/main/java/com/umbrella/insurance/core/models/entities/PendingPolicy.java
package com.umbrella.insurance.core.models.entities;

import com.umbrella.insurance.core.deserializers.DateDeserializer;
import com.umbrella.insurance.core.deserializers.TimestampDeserializer;
import com.umbrella.insurance.core.serializers.DateSerializer;
import com.umbrella.insurance.core.serializers.TimestampSerializer;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.sql.Date;
import java.sql.Timestamp;
import java.util.List;

@Getter
@Setter
@Entity
@Table(name = "pending_policies", schema = "public", uniqueConstraints = {
        @UniqueConstraint(name = "pending_policies_unique", columnNames = {"pending_policy_name"})
})
public class PendingPolicy {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "pending_policy_id", nullable = false)
    private Long id;

    @Column(name = "premium_amount")
    private Double premiumAmount;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "location_id")
    private Location location;

    @Column(name = "coverage_amount")
    private Double coverageAmount;

    @Column(name = "implied_probability")
    private Double impliedProbability;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "unit_id")
    private Unit unit;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "peril_id")
    private Peril peril;

    @JsonSerialize(using = DateSerializer.class)
    @JsonDeserialize(using = DateDeserializer.class)
    @NotNull
    @Column(name = "start_date", nullable = false)
    private Date startDate;

    @JsonSerialize(using = DateSerializer.class)
    @JsonDeserialize(using = DateDeserializer.class)
    @NotNull
    @Column(name = "end_date", nullable = false)
    private Date endDate;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "pending_policy_state_id")
    private PendingPolicyState pendingPolicyState;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "session_id")
    private Session session;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "fee_id")
    private Fee fee;

    @NotNull
    @Column(name = "pending_policy_name", nullable = false, length = Integer.MAX_VALUE)
    private String pendingPolicyName;

    @OneToMany(mappedBy = "originalPendingPolicy", cascade = CascadeType.ALL)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    //@JoinColumn(name = "split_pending_policy_id_1")
    private List<PendingPolicy> splitPendingPolicy1;

    @OneToMany(mappedBy = "originalPendingPolicy", cascade = CascadeType.ALL)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    //@JoinColumn(name = "split_pending_policy_id_2")
    private List<PendingPolicy> splitPendingPolicy2;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "original_pending_policy_id")
    private PendingPolicy originalPendingPolicy;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "account_balance_escrow_transaction_id")
    private AccountBalanceTransaction accountBalanceEscrowTransaction;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "account_balance_canceled_escrow_transaction_id")
    private AccountBalanceTransaction accountBalanceCanceledEscrowTransaction;

    @JsonSerialize(using = TimestampSerializer.class)
    @JsonDeserialize(using = TimestampDeserializer.class)
    @NotNull
    @Column(name = "created_date_time", nullable = false)
    private Timestamp createdDateTime;

    @JsonSerialize(using = TimestampSerializer.class)
    @JsonDeserialize(using = TimestampDeserializer.class)
    @Column(name = "canceled_date_time")
    private Timestamp canceledDateTime;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "pending_policy_type_id")
    private PendingPolicyType pendingPolicyType;

    @ManyToOne(fetch = FetchType.EAGER)
    @OnDelete(action = OnDeleteAction.SET_NULL)
    @JoinColumn(name = "order_type_id")
    private OrderType orderType;

}